allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -271,6 +310,166 @@ def forward(self,  # type: ignore
+            entity2neighbors = world.table_graph.neighbors
+            neighbor_indexes = []
+            for entity in entities:
+                entity_neighbors = [entity2indices[cn] for cn in entity2neighbors[entity]]
+                padded = pad_sequence_to_length(entity_neighbors, num_entities)
+                neighbor_indexes.append(padded)
+            neighbor_indexes = pad_sequence_to_length(neighbor_indexes,
+                                                      num_entities,
+                                                      lambda: [0] * num_entities)
+            batches_neighbors.append(neighbor_indexes)
+        return Variable(tensor.data.new(batches_neighbors)).long()
+
+    @staticmethod
+    def _get_type_vector(worlds: List[WikiTablesWorld],
+                         num_entities: int,
+                         tensor: torch.Tensor) -> Tuple[Variable, Dict[int, int]]:
+        """
+        Produces the one hot encoding for each entity's type. In addition,
+        a map from index to type is returned to combine entity type operations
+        into one method.
+
+        Parameters
+        ----------
+        worlds : ``List[WikiTablesWorld]``
+        num_entities : ``int``
+        tensor : ``torch.Tensor``
+            Used for copying the constructed list onto the right device.
+
+        Returns
+        -------
+        A ``torch.autograd.Variable`` with shape ``(batch_size, num_entities, num_types)``.
+        entity_types : ``Dict[int,str]``
+            This is a mapping from (batch_index, entity_index) to entity type id.
+        """
+        entity_types = {}
+        batch_types = []
+        for batch_index, world in enumerate(worlds):
+            types = []
+            for entity_index, entity in enumerate(world.table_graph.entities):
+                cell_type_one_hot = [1, 0]
+                row_type_one_hot = [0, 1]
+                types.append((cell_type_one_hot if entity.startswith('fb:cell') else row_type_one_hot))
+
+                # For easier lookups later, we're actually using a _flattened_ version
+                # of (batch_index, entity_index) for the key, because this is how the
+                # linking scores are stored.
+                flattened_entity_index = batch_index * num_entities + entity_index
+                entity_type = 0 if entity.startswith('fb:cell') else 1
+                entity_types[flattened_entity_index] = entity_type
+            padded = pad_sequence_to_length(types, num_entities, lambda: [0, 0])
+            batch_types.append(padded)
+        return Variable(tensor.data.new(batch_types)), entity_types
+
+    def _get_prob_entity(self,
+                         worlds: List[WikiTablesWorld],
+                         scores: Variable,
+                         batch_size: int,
+                         num_entities: int,
+                         num_question_tokens: int,
+                         tensor: torch.Tensor) -> Variable:
```

Comment:
This needs to take into account the question mask and the entity mask.

Reply:
Good point, I added the question mask through a masked_softmax. How do I incorporate the entity mask? Isn't it already accounted for when the index select picks valid entities of each type?

Reply:
`index_select` gets passed a padded tensor, I believe.  It'd be a lot easier to answer this question after seeing a concrete test case for the method.